writer_test.go

```diff
@@ -26,6 +30,61 @@ func TestWriteAndRetryFails(t *testing.T) {
 	}
 }
 
+// Handle cases where a remote server hangs up (e.g. idle timeout),
+// without dropping logs.
+func TestSocketClose(t *testing.T) {
+	// use ioutil.TempFile to get a name that is unique
+	f, err := ioutil.TempFile("", "syslogtest")
+	if err != nil {
+		t.Errorf("failed creating temp file: %v", err)
+		return
+	}
+	f.Close()
+	la := f.Name()
+	os.Remove(la)
+
+	l, err := net.Listen("unix", la)
+	if err != nil {
+		t.Errorf("failed starting server: %v", err)
+		return
+	}
+
+	// This server closes the client connection after receiving one packet,
+	// and then listens for a second connection.
+	//
+	// We want to ensure that our writer reconnects after the socket is closed.
+	go func(listener *net.UnixListener) {
+		c, err := listener.Accept()
+		b := make([]byte, 256)
+		c.Read(b)
+		c.Close()
+
+		listener.SetDeadline(time.Now().Add(50 * time.Millisecond))
+		c, err = listener.Accept()
+		if err != nil {
+			t.Errorf("didn't get a reconnection: %v", err)
+			return
+		}
+
+		b = make([]byte, 256)
+		c.Read(b)
+		c.Close()
+		l.Close()
```

Comment:
I believe we want this to be `listener.Close()`, because `l` doesn't belong to this scope. It probably doesn't matter for _this_ test, but it's more correct.